sa/storage-authority.go

```diff
@@ -673,6 +674,117 @@ func (ssa *SQLStorageAuthority) FinalizeAuthorization(authz core.Authorization)
 	return
 }
 
+var authorizationTables = []string{
+	"authz",
+	"pendingAuthorizations",
+}
+
+const getAuthorizationIDsMax = 1000
+
+func getAuthorizationIDsByDomain(db *gorp.DbMap, ident string, now time.Time) ([]string, error) {
+	var allIDs []string
+	for _, tableName := range authorizationTables {
+		_, err := db.Select(
+			&allIDs,
+			fmt.Sprintf(
+				`SELECT id FROM %s
+         WHERE identifier = :ident AND
+         status != :invalid AND
+         status != :revoked AND
+         expires > :now
+         LIMIT :limit`,
+				tableName,
+			),
+			map[string]interface{}{
+				"ident":   ident,
+				"invalid": string(core.StatusInvalid),
+				"revoked": string(core.StatusRevoked),
+				"now":     now,
+				"limit":   getAuthorizationIDsMax,
+			},
+		)
+		if err != nil {
+			return nil, err
+		}
+		if len(allIDs) == getAuthorizationIDsMax {
+			break
+		}
+	}
+	return allIDs, nil
+}
+
+func revokeAuthorizations(db *gorp.DbMap, authIDs []string) (int64, int64, error) {
+	results := []int64{0, 0}
+	quotedIDs := []string{}
+	for _, id := range authIDs {
+		quotedIDs = append(quotedIDs, strconv.Quote(id))
+	}
+	idStr := strings.Join(quotedIDs, ", ")
+	for i, tableName := range authorizationTables {
+		result, err := db.Exec(
+			fmt.Sprintf(
+				`UPDATE %s
+         SET status = ?
+         WHERE id IN (%s)`,
+				tableName,
+				idStr,
+			),
+			string(core.StatusRevoked),
+		)
+		if err != nil {
+			return results[0], results[1], err
+		}
+		batchSize, err := result.RowsAffected()
+		if err != nil {
+			return results[0], results[1], err
+		}
+		results[i] = batchSize
+	}
+	_, err := db.Exec(
+		fmt.Sprintf(
+			`UPDATE challenges
```

Comment:
> Larger question: why do we need revoke the challenges? I know there's got to be a reason but I've forgotten it.

We don't super need to, I think we just felt like it would be tidier. The check for "pending" status when POSTing to a challenge checks the status of the authorization, not the challenge. In the future we may want to check the challenge status as well.

So we could say that revoking the authorization is sufficient, and document clearly in UpdateAuthorizations that it's important to check the authorization status even if we're checking challenge status too.